chalk-derive/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use syn::{parse_macro_input, Data, DeriveInput, GenericParam, Ident, TypeParamBound};
+
+#[proc_macro_derive(Fold, attributes(has_type_family))]
+pub fn derive_fold(item: TokenStream) -> TokenStream {
```

Comment:
I would like to see some documentation on the conventions being observed here (i.e., the fact that one should have a `TF: TypeFamily` attribute or whatever).